### DISCUSSION TITLE
How to pass data to post request for testing.

### DISCUSSION THREAD

**Discussion by Fahad-Md-Kamal** [fastapi/fastapi#9162] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# ROUTE:

@router.post('/token', response_model=schemas.Token)
async def login(user_credentials: OAuth2PasswordRequestForm = Depends(), db: Session = Depends(get_db)):
    """
    Login to the system with (Email | Username | Contact)
    """

    user = db.query(models.User).filter(
        (models.User.email == user_credentials.username) |
        (models.User.username == user_credentials.username) |
        (models.User.contact == user_credentials.username)
    ).first()

    if not user:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN, detail="Invalid Credentials"
        )

    if not utils.verify_pwd(user_credentials.password, user.password):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN, detail="Invalid Credentials"
        )

    access_token = oauth2.create_access_token(data={'user_id': user.id})

    return {"access_token": access_token, "token_type": "bearer"}


# Test Code: 


from fastapi.testclient import TestClient

from ..main import app


client = TestClient(app)


def test_auth_token(get_client: TestClient):
    client = get_client.post('/token', json={"username": "admin", "password": "1234567890"})
    assert client.status_code == 200
```

### Description

I'm trying to test the login endpoint with required payload username and password. But when I pass the data and submit the form with post request I get **422 Unprocessable Entity** error.

It says: 
```bash
{'detail': [{'loc': ['body', 'username'], 'msg': 'field required', 'type': 'value_error.missing'}, {'loc': ['body', 'password'], 'msg': 'field required', 'type': 'value_error.missing'}]}
```

What I'm missing here and how should I submit payload for post api's.

### Operating System

macOS

### FastAPI Version

0.89.1

### Python Version

Python 3.10.8

**Comment by jgould22** [comment, marked as answer]
I suspect OAuth2PasswordRequestForm wants the post to be form-data(based on the name) and you have supplied it json.

You need to post it as form data. Try it as described here and see if that works 

https://www.python-httpx.org/quickstart/#sending-form-encoded-data

**Reply by Fahad-Md-Kamal** [reply]
thanks
